presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/RuleIndex.java

```diff
@@ -71,6 +71,14 @@ public Builder register(Rule matchable)
             return this;
         }
 
+        private Pattern<?> getFirstPattern(Pattern<?> pattern)
```

Comment:
I'm not convinced. It's specific to RuleIndex. Even if there were more usages I would probably prefer a PatternUtil, PatternOperations, or sth like that. Would rather not extend `Pattern`'s interface till its key to its function.